crates/data_components/src/s3_vectors/partition.rs

```diff
@@ -0,0 +1,231 @@
+    pub fn to_index_name(&self) -> String {
+        [
+            self.index_name.clone(),
+            self.column_name.clone(),
+            self.partition_value_hash.clone(),
+            self.partition_by_hash.clone(),
+        ]
+        .join(PARTS_SEPARATOR)
+    }
+
+    pub fn from_index_name(index_name: &str) -> Result<Self, Error> {
+        let parts: Vec<&str> = index_name.split('-').collect();
+        let num_parts = parts.len();
+        ensure!(num_parts == 4, IncorrectNumPartsInNameSnafu { num_parts });
+        Ok(Self {
+            index_name: parts[0].to_string(),
+            column_name: parts[1].to_string(),
+            partition_value_hash: parts[2].to_string(),
+            partition_by_hash: parts[3].to_string(),
+        })
+    }
+
+    /// Determines if the partitions come from the same dataset
+    #[must_use]
+    pub fn belongs_with(&self, other: &Self) -> BelongsWith {
+        if self.index_name != other.index_name || self.column_name != other.column_name {
+            return BelongsWith::DifferentDataset;
+        }
+        if self.partition_by_hash != other.partition_by_hash {
+            return BelongsWith::DifferentParitionByExpressions;
+        }
+        BelongsWith::SameDataset
+    }
+}
+
+fn sanitize(s: &str) -> String {
+    s.replace('_', "-")
+}
+
+fn truncate(s: &str, len: usize) -> String {
+    s.chars().take(len).collect()
+}
+
+fn hash_to_hex(input: &str) -> String {
+    let hash = XxHash64::oneshot(HASH_SEED, input.as_bytes());
+    format!("{hash:x}")
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use datafusion::prelude::col;
+    use datafusion::scalar::ScalarValue;
+
+    /// See [CreateIndex](https://docs.aws.amazon.com/AmazonS3/latest/API/API_S3VectorBuckets_CreateIndex.html#API_S3VectorBuckets_CreateIndex_RequestSyntax)
+    const S3_VECTOR_INDEX_NAME_MAX_LENGTH: usize = 63;
+
+    #[test]
+    fn index_name_length_restricted() {
+        let index_name = "a".repeat(INDEX_NAME_MAX_LENGTH + 1);
+        let column_name = "col1";
+        let partition_value = ScalarValue::from("val");
+        let partition_by = vec![col("col1")];
+
+        let result =
+            PartitionedIndexName::new(&index_name, column_name, &partition_value, &partition_by)
+                .expect("result");
+        let result = result.to_index_name();
+
+        assert!(result.len() <= S3_VECTOR_INDEX_NAME_MAX_LENGTH);
+        assert_eq!(
+            result.split('-').next().expect("next").len(),
+            INDEX_NAME_MAX_LENGTH
+        );
+    }
+
+    #[test]
+    fn new_index_partition_name() {
+        let index_name = "test_index";
+        let column_name = "test_col";
+        let partition_value = ScalarValue::from("value");
+        let partition_by = vec![col("col1")];
+
+        let result =
+            PartitionedIndexName::new(index_name, column_name, &partition_value, &partition_by)
+                .expect("new");
+
+        assert_eq!(result.index_name, "test-index");
```

Comment:
One of the side effects of the sanitization is that the index name is different from what is passed in - is that going to be a problem in the calling code?

Reply:
Or maybe a simpler solution is instead of sanitizing, just require that the index_name/column_name not have  `-` or `_` in them. That might work if we control the `index_name`/`column_name` - but if that comes from user input where then we would need to maintain a mapping in the calling code.

Reply:
I don't think so. The comparison in `belongs_with` happens after sanitizing.

Although, we should just error if the index name contain a hyphen because they just can't name indexes with hyphens due to AWS restrictions.